### PR TITLE
Single WG implementation of `__parallel_find_or`

### PR DESCRIPTION
**Pull request by SergeyKopienko**
In this PR we prepare single work-group implementation of `__parallel_find_or` :
- they doesn't used `atomic`-based synchronization;
- they doesn't used `syc::buffer` for return result when USM-memory is available on device (we using `__result_and_scratch_storage` prepared by @julianmi earlier);
- kernel's compilation has been removed from `__parallel_find_or` and their staff.

This approach gives us a big performance boost for small data sizes.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1153]: Why don't we remove this throughout the single WG implementation? The compiler will probably optimize this out, but it would shorten the code.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1155]: Should we move this comment to the main routine to make clear how the single WG implementation works? This would probably follow the reader.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1159]: This looks misplaced and you have a `main parallel_for` comment below. I would remove this.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1205]: These look outdated. I would remove them.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1197]: This comment looks outdated. I would remove it.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1211]: This comment looks outdated. I would remove it.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1218]: Why do we create a future here? Can't we just wait on the event and call `__get_value` on `__result_storage`?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1168]: `__event_id` suggests to me that this is the event's index instead of the event. I would simply name it `__event`.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1241]: Same comment about removing this since I don't think it adds value to the reader being placed right in the beginning.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1339]: Extending my comment about compiling the kernels: I would be interested to see if we even need it for the multi WG implementation since we don't use shared resources.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1355]: Should we move this done to line 1371 for brevity?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1323]: I would remove this and similar comments since they weren't there before and these are typical patterns in our code.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1329]: Same about removing this comment.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1348]: Same about removing this comment.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1363]: Same about removing this comment.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1379]: I would remove this comment. I think it's very clear at this point. Instead, I would add some comments about the single and multi WG implementations above and below.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Removed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Removed.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Removed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Removed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Removed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Removed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Removed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Removed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1192]: Why?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1087]: Perf looks equal, this change has been rolled back.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1192]: I misread the comment in line 1197. I would still suggest removing this line here and in the multi WG kernel since we explain that these are group operations already in line 1194 and 1196.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1192]: Is it really significant?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Who will call `sycl::queue::wait()` method in this case?

Comment (julianmi) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: The caller of `__get_value` needs to ensure that the result is ready. See the comment in https://github.com/oneapi-src/oneDPL/blob/825ba2f02fe8110e93afa1a49c7292d646851690/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h#L628-L629.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Kernel's compilation for `__parallel_find_or` and their staff has been removed.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: How our __future::get() is implemented:
```C++
    auto
    get()
    {
        if constexpr (sizeof...(_Args) > 0)
        {
            auto& __val = std::get<0>(*this);
            return __wait_and_get_value(__val);
        }
        else
            wait();
    }
```
 
How __wait_and_get_value is implemented:
```C++
    template <typename _ExecutionPolicy, typename _T>
    constexpr auto
    __wait_and_get_value(__result_and_scratch_storage<_ExecutionPolicy, _T>& __storage)
    {
        if (__storage.is_USM())
            wait();
        return __storage.__get_value();
    }
```
 
So for me this mean that we really should call wait() only for USM memory.

@julianmi, am I correct?

Comment (julianmi) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Yes, we only need explicit synchronization when using USM. I think we should remove the future since it is not used and change this line to `if (__result_storage.is_USM()) wait(); __result_storage.__get_value();`.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1174]: No but it would shorten the code. Your decision.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1187]: I don't know if this works if `__result_and_scratch_storage_t` is in buffer storage mode.  This would return an accessor, which doesn't overload the `operator*()`, I believe.

I think the better way is to not dereference, but rather assign to an `auto` typed variable, then assign `__found_local` to the zeroth element of the accessor / usm ptr.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1140]: Fixed: renamed to `__FoundStateType`

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: As I may see below, the function `__get_usm_or_buffer_accessor_ptr` always return pointer:
```C++
    template <typename _Acc>
    static auto
    __get_usm_or_buffer_accessor_ptr(const _Acc& __acc, ::std::size_t __scratch_n = 0)
    {
#if _ONEDPL_SYCL_UNIFIED_USM_BUFFER_PRESENT
        return __acc.__get_pointer();
#else
        return &__acc[__scratch_n];
#endif
    }
```

So we simple apply `operator*` to this pointer and save result into reference:
```C++
                    __FoundStateType& __found =
                        *__result_and_scratch_storage_t::__get_usm_or_buffer_accessor_ptr(__result_acc);
```

Looks like this code hasn't problems here...

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1174]: Let's save as is.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: I would suggest adding a `__result_and_scratch_storage.__wait_and_get_value` with this logic and use it here and in `__future.__wait_and_get_value` to not duplicate code.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Should this new implementation wait under `ONEDPL_ALLOW_DEFERRED_WAITING` macro check like in the `__future` struct?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: I agree with Dan. Assigning the value to the zeroth element seems the safer choice here and is used in our existing code already. If you want to stick with this code, please test it with the buffer mode by setting `_ONEDPL_SYCL_UNIFIED_USM_BUFFER_PRESENT` to false or use an older compiler. Another path you would need to try is setting `__use_USM_host` and `__supports_USM_device` to false in `__result_and_scratch_storage`.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: > I would suggest adding a `__result_and_scratch_storage.__wait_and_get_value` with this logic and use it here and in `__future.__wait_and_get_value` to not duplicate code.

done.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: `sycl::accessor` based on `sycl::buffer` return in this case `reference` type.
After that inside the method
```C++
    auto
    __get_pointer() const // should be cached within a kernel
    {
        return __usm ? __ptr + __offset : &__acc[__offset];
    }
```
we return address from this reference.
So I think it's look ok too.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: About the other our cases, for example `__work_group_reduce_kernel` - there is double referencing: we pass pointer into `__work_group_reduce_kernel` by const reference, names it as `__res_acc` and then, as you wrote, assign value to in through `[0]'. For me this practice isn't very good.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed: I changed this place for compatibility with other code, but not because it's incorrect.

**Comment by SergeyKopienko** [comment]
@julianmi , @danhoeflinger , @adamfidel have somebody additional comments for this PR? Looks like all comments has been fixed now.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Ah, I guess I misunderstood how `__get_usm_or_buffer_accessor_ptr()` worked, but your logic is correct I think @SergeyKopienko . 

The naming of the function gives it away I suppose, a pointer to a buffer accessor.   Its just not a normal thing to do, but it does allow the return to be treated equivalently to the usm pointer, I guess.

I think its more clear as you have it written now, but I'm OK with either option.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1304]: Updated comment https://github.com/oneapi-src/oneDPL/pull/1736#discussion_r1697050711. Should we remove this and integrate the if-statement into line 1307 for brevity?

**Review by danhoeflinger — DISMISSED** [review]
LGTM, please come to some agreement with @julianmi on his comment before merging, I don't have a strong preference there.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1304]: I prefer to save this as is because we should have two separate steps
1) tune amount of groups
2) run one of two implementations.
I prefer to have separate steps and do not mix them in one condition.

Comment (julianmi) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: We discussed to move this to where we define the launch configuration based on Sergey's preference.

**Review by julianmi — APPROVED** [review]
LGTM. Please wait for green CI.

**Pull request merged**
